lxd/cluster/membership.go

```diff
@@ -24,6 +24,11 @@ import (
 	"github.com/canonical/lxd/shared/version"
 )
 
+// clusterBusyError is returned by dqlite if attempting attempting to join a cluster at the same time as a role-change.
+// This error tells us we can retry and probably join the cluster or fail due to something else.
+// The error code here is sqlite3.ErrBusy.
+var clusterBusyError = fmt.Errorf("a configuration change is already in progress (5)")
```

Comment:
Is this a go-dqlite or underlying dqlite error?
If so we sglukd use type detection (possibly with dqlite error code matching) rather than string matching.

There's an example of this in lxd already I'll dig out for you.

Reply:
It's from dqlite.

Reply:
Well, if it helps its a `protocol.ErrRequest` from `go-dqlite` but the error message and code are from `dqlite`

Reply:
Something like this:

```go
// Insert a new Storage Bucket record.
		result, err := tx.tx.Exec(`
		INSERT INTO storage_buckets
		(storage_pool_id, node_id, name, description, project_id)
		VALUES (?, ?, ?, ?, (SELECT id FROM projects WHERE name = ?))
		`, poolID, nodeID, info.Name, info.Description, projectName)
		if err != nil {
			var dqliteErr dqliteDriver.Error
			// Detect SQLITE_CONSTRAINT_UNIQUE (2067) errors.
			if errors.As(err, &dqliteErr) && dqliteErr.Code == 2067 {
				return api.StatusErrorf(http.StatusConflict, "A bucket for that name already exists")
			}

			return err
		}
```

From https://github.com/canonical/lxd/blob/main/lxd/db/storage_buckets.go#L304-L312

Reply:
Your hint led me to find https://github.com/canonical/lxd/blob/fdc3dc68d9ed4b71d2676966cd10814ec2cf3638/lxd/db/query/retry.go#L55-L88 which looks like it handles several retriable errors from dqlite, including the one we want in this case`ErrBusy`.

Reply:
@tomponline Spoke too soon. this error's actually not a `driver.Error`, it's a `protocol.ErrRequest` which isn't exposed externally in `go-dqlite`...

Reply:
Drafted it so you wouldn't accidentally merge it :)

Reply:
Not till tesrs run :)

Reply:
I wonder if @MathieuBordere or @cole-miller could help with accessing this error.

Reply:
Ok so the issue is that packages from https://github.com/canonical/go-dqlite/blob/4dfb00574f2633c7153e4d8e622016510448f4c0/driver/driver.go#L812 will be cast as a `driver.Error` but in this case we're dealing with an error from a client function here https://github.com/canonical/go-dqlite/blob/4dfb00574f2633c7153e4d8e622016510448f4c0/client/client.go#L171 which is left as the internal `protocol.ErrRequest` type defined here https://github.com/canonical/go-dqlite/blob/4dfb00574f2633c7153e4d8e622016510448f4c0/internal/protocol/errors.go#L18

Reply:
Reverting this to use the hard-coded error message for now.

Reply:
Sorry for the delay in responding. Could we make this more convenient for LXD by making sure a `driver.Error` is returned here?